src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -24,52 +25,36 @@ public ITransaction CreateTransaction(ISessionImplementor session)
 
 		public void EnlistInDistributedTransactionIfNeeded(ISessionImplementor session)
 		{
+			// Ensure the session does not run on a thread supposed to be blocked, waiting
+			// for transaction completion.
+			session.TransactionContext?.WaitOne();
```

Comment:
Transaction completion may concurrently nullify it, but `?.` handles that (correctly evaluate it only once, guaranteeing it will not try to call `WaitOne` on null).